services/oauth/v2/utils.go

```diff
@@ -0,0 +1,141 @@
+package v2
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"strconv"
+	"syscall"
+	"time"
+
+	jsoniter "github.com/json-iterator/go"
+	"github.com/tidwall/gjson"
+
+	"github.com/rudderlabs/rudder-go-kit/stats"
+	router_utils "github.com/rudderlabs/rudder-server/router/utils"
+	"github.com/rudderlabs/rudder-server/utils/misc"
+)
+
+var jsonfast = jsoniter.ConfigCompatibleWithStandardLibrary
+
+const (
+	REFRESH_TOKEN = "REFRESH_TOKEN"
+	// Identifier to be sent from destination(during transformation/delivery)
+	AUTH_STATUS_INACTIVE = "AUTH_STATUS_INACTIVE"
+	// Identifier for invalid_grant or access_denied errors(during refreshing the token)
+	REF_TOKEN_INVALID_GRANT = "ref_token_invalid_grant"
+)
+
+var ErrPermissionOrTokenRevoked = errors.New("Problem with user permission or access/refresh token have been revoked")
+
+func GetOAuthActionStatName(stat string) string {
+	return fmt.Sprintf("oauth_action_%v", stat)
+}
+
+func (authStats *OAuthStats) SendTimerStats(startTime time.Time) {
+	statsTags := stats.Tags{
+		"id":              authStats.id,
+		"workspaceId":     authStats.workspaceId,
+		"rudderCategory":  authStats.rudderCategory,
+		"isCallToCpApi":   strconv.FormatBool(authStats.isCallToCpApi),
+		"authErrCategory": authStats.authErrCategory,
+		"destType":        authStats.destDefName,
+		"flowType":        string(authStats.flowType),
+		"action":          authStats.action,
+		"oauthVersion":    "v2",
+	}
+	stats.Default.NewTaggedStat(authStats.statName, stats.TimerType, statsTags).SendTiming(time.Since(startTime))
+}
+
+// Send count type stats related to OAuth(Destination)
+func (refStats *OAuthStats) SendCountStat() {
+	statsTags := stats.Tags{
+		"oauthVersion":    "v2",
+		"id":              refStats.id,
+		"workspaceId":     refStats.workspaceId,
+		"rudderCategory":  refStats.rudderCategory,
+		"errorMessage":    refStats.errorMessage,
+		"isCallToCpApi":   strconv.FormatBool(refStats.isCallToCpApi),
+		"authErrCategory": refStats.authErrCategory,
+		"destType":        refStats.destDefName,
+		"isTokenFetch":    strconv.FormatBool(refStats.isTokenFetch),
+		"flowType":        string(refStats.flowType),
+		"action":          refStats.action,
+	}
+	stats.Default.NewTaggedStat(refStats.statName, stats.CountType, statsTags).Increment()
+}
```

Comment:
Receiver names should be same and short.
```suggestion
func (s *OAuthStats) SendTimerStats(startTime time.Time) {
	statsTags := stats.Tags{
		"id":              s.id,
		"workspaceId":     s.workspaceId,
		"rudderCategory":  s.rudderCategory,
		"isCallToCpApi":   strconv.FormatBool(s.isCallToCpApi),
		"authErrCategory": s.authErrCategory,
		"destType":        s.destDefName,
		"flowType":        string(s.flowType),
		"action":          s.action,
		"oauthVersion":    "v2",
	}
	stats.Default.NewTaggedStat(s.statName, stats.TimerType, statsTags).SendTiming(time.Since(startTime))
}

// Send count type stats related to OAuth(Destination)
func (s *OAuthStats) SendCountStat() {
	statsTags := stats.Tags{
		"oauthVersion":    "v2",
		"id":              s.id,
		"workspaceId":     s.workspaceId,
		"rudderCategory":  s.rudderCategory,
		"errorMessage":    s.errorMessage,
		"isCallToCpApi":   strconv.FormatBool(s.isCallToCpApi),
		"authErrCategory": s.authErrCategory,
		"destType":        s.destDefName,
		"isTokenFetch":    strconv.FormatBool(s.isTokenFetch),
		"flowType":        string(s.flowType),
		"action":          s.action,
	}
	stats.Default.NewTaggedStat(s.statName, stats.CountType, statsTags).Increment()
}
```